### PR TITLE
cppcheck/*: Test against dummy file fix #5831

### LINKED ISSUE #5831: [package] cppcheck/2.4.1: Running cppcheck missing cfg files
### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **cppcheck/2.4.1**
  * Operating System+version: **Linux Ubuntu 18.04**
  * Compiler+version: **GCC 7.5**
  * Conan version: **conan 1.37.1**
  * Python version: **Python 3.6.9**

### Conan profile (output of `conan profile show default` or `conan profile show <profile>` if custom profile is in use)
```
Configuration for profile default:

[settings]
os=Linux
os_build=Linux
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=7
compiler.libcxx=libstdc++11
build_type=Release
[options]
[build_requires]
[env]
```

### Steps to reproduce (Include if Applicable)

* Install cppcheck with a runenv:
```
$ conan install "cppcheck/2.4.1@" -r conan-center  -g virtualrunenv
$ source activate_run.sh
```
* Run cppcheck on directory containing C++ files
```
cppcheck --xml --enable=warning,style,performance --std=c++11 . 2> test.xml
cppcheck: Failed to load library configuration file 'std.cfg'. File not found
```

It seems that cppcheck needs to be build locally because it uses some hardcoded paths, so `build_policy = "always"` should be set.
Also [conanfile.py#L66](https://github.com/conan-io/conan-center-index/blob/master/recipes/cppcheck/all/conanfile.py#L66) deletes the `share` directory that's where the *cfg* files are installed by default and where the hardcoded path points to.
With this line deleted (and build locally) I can run the cppcheck installed via conan properly.

If you want, I can prepare a PR for these 2 changes.

### PR DESCRIPTION
**Pull request by MartinDelille**
Specify library name and version:  **cppcheck/...**

This test hightlights an issue in the current cppcheck recipe which fails to detect std.cfg.
Any idea how to fix this ?

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by MartinDelille** [comment]
I ping the previous contributors to this recipe: @mjvankampen @SpaceIm @Pro @prince-chrismc @ericLemanissier

**Comment by ericLemanissier** [comment]
ok, so AFAIU, std.cfg is loaded in [CppCheckExecutor::check_internal](https://github.com/danmar/cppcheck/blob/8f386e15fdedff37486c683d933ccc9a1e307388/cli/cppcheckexecutor.cpp#L873) which calls [CppCheckExecutor::tryLoadLibrary](https://github.com/danmar/cppcheck/blob/8f386e15fdedff37486c683d933ccc9a1e307388/cli/cppcheckexecutor.cpp#L1135)
which calls [Library::load](https://github.com/danmar/cppcheck/blob/da1e2b22be7afb2918994d97c5071e6cdc1382dd/lib/library.cpp#L103-L110)

it first searches for `std.cfg` in the same folder as the cppcheck executable, then in the `cfg` subfolder next to the cppcheck executable, and If `FILESDIR` is set during the build, it finally tries in `FILESDIR/cfg`.

All in all, FILESDIR is always used as an absolute path, so it does not seem to be a good fit for conan. 

By the way, [I tested the current recipe](https://github.com/eirikb/proof-of-conan/runs/6127037166?check_suite_focus=true), and it works for:
- all versions of cppcheck on windows 
- cppcheck>= 2.7.4 on linux

so it fails for:
- cppcheck <= 2.7.3 on linux
- all versions of cpcheck on macos

I think the best course is to report the macos issue upstream (https://sourceforge.net/p/cppcheck/discussion/ I think)

**Comment by MartinDelille** [comment]
> All in all, FILESDIR is always used as an absolute path, so it does not seem to be a good fit for conan.

I don't understand what could be a good fit for conan then.

I opened a thread here: https://sourceforge.net/p/cppcheck/discussion/general/thread/1ad82367da/

**Comment by MartinDelille** [comment]
@ericLemanissier What is

File: recipes/cppcheck/all/conanfile.py [hunk 1]
```diff
@@ -44,16 +43,16 @@ def source(self):
...
-        self._cmake.definitions["USE_MATCHCOMPILER"] = "Auto"
-        self._cmake.definitions["ENABLE_OSS_FUZZ"] = False
-        self._cmake.configure(build_folder=self._build_subfolder)
-        return self._cmake
+        cmake = CMake(self)
+        cmake.definitions["USE_Z3"] = self.options.with_z3
+        cmake.definitions["HAVE_RULES"] = self.options.have_rules
+        cmake.definitions["USE_MATCHCOMPILER"] = "Auto"
+        cmake.definitions["ENABLE_OSS_FUZZ"] = False
+        cmake.definitions["FILESDIR"] = os.path.join(self.package_folder, "bin")
```
Comment (MartinDelille) [inline recipes/cppcheck/all/conanfile.py:53]: @ericLemanissier I don't understand why setting FILESDIR wouldn't be "a good fit for conan" ?

Comment (MartinDelille) [reply recipes/cppcheck/all/conanfile.py:53]: (I tested it locally on my Mac and it works)

Comment (MartinDelille) [reply recipes/cppcheck/all/conanfile.py:53]: Ok I think I understand: if I get the binary from conan center, FILESDIR will be an absolute path on the build machine and won't be valid locally.

Comment (madebr) [reply recipes/cppcheck/all/conanfile.py:54]: I grepped in the current master of cppcheck for `#ifdef FILESDIR` and found 9 matches.
I think every occurrence needs to be changed to call a helper functions that does the following:
```c++
const char* filesdir(const char* default) {
#ifdef FILESDIR
    return FILESDIR
#else
    if (::getenv("CPPCHECK_FILESDIR")) {
        return ::getenv("CPPCHECK_FILESDIR");
    }
    return default;
#endif
}
```

Imho, best would be to create such a patch for current master and send it upstream for feedback + telling why you want to add it.
Based on the feedback, you can improve the patch + add the patch here.

**Comment by SSE4** [comment]
okay, the error is:
```
ERROR: Conanfile 'cppcheck/2.7.3' has build_policy='always', no packages can be uploaded
```

why is it needed? https://github.com/conan-io/conan-center-index/pull/10423/files#diff-e0631df37f64c1c2264732cecd7fbb4f60c6cfbe97de586e33212857eed10639R20

**Comment by madebr** [comment]
> okay, the error is:
> 
> ```
> ERROR: Conanfile 'cppcheck/2.7.3' has build_policy='always', no packages can be uploaded
> ```
> 
> why is it needed? https://github.com/conan-io/conan-center-index/pull/10423/files#diff-e0631df37f64c1c2264732cecd7fbb4f60c6cfbe97de586e33212857eed10639R20

I think because `FILESDIR` will embed a path of the build machine into the executable, making cppcheck non-portable.

File: recipes/cppcheck/all/conanfile.py (hunk 1 above)
Comment (ericLemanissier) [reply recipes/cppcheck/all/conanfile.py:54]: > Another possibility, but only applicable to portable applications (not for applications aimed for linux distributions) would be to get the path of the running executable and build a path on top of that.\nBut such code would be platform dependent.

AFAIU, this is already what cppcheck does when FILESDIR macro is not set, except that it does not work on macos for some reason

Comment (madebr) [reply recipes/cppcheck/all/conanfile.py:54]: It does things like the following:
```c++
void Settings::loadCppcheckCfg()
{
    std::string fileName = Path::getPathFromFilename(exename) + "cppcheck.cfg";
#ifdef FILESDIR
    if (Path::fileExists(FILESDIR "/cppcheck.cfg"))
        fileName = FILESDIR "/cppcheck.cfg";
#endif

    std::ifstream fin(fileName);
   // ... (do actual work)
```
and `getPathFromFilename` does the following:
```
std::string Path::getPathFromFilename(const std::string &filename)
{
    const std::size_t pos = filename.find_last_of("\\/");

    if (pos != std::string::npos)
        return filename.substr(0, 1 + pos);

    return "";
}
```

So it `cppcheck.cfg` will be found when you run `cppcheck` as e.g. `~/.conan/data/cppcheck/1.2.3.4/_/_/package/123456/bin/cppcheck`, but not when you run it as `cppcheck` (assuming you have `PATH` set-up).

Comment (madebr) [reply recipes/cppcheck/all/conanfile.py:54]: `exename` is initialized as follows on Windows:

```c++
#ifdef _WIN32
#include <windows.h>

static char exename[1024] = {0};
#endif

/**
 * Main function of cppcheck
 *
 * @param argc Passed to CppCheck::parseFromArgs()
 * @param argv Passed to CppCheck::parseFromArgs()
 * @return What CppCheckExecutor::check() returns.
 */
int main(int argc, char* argv[])
{
    // MS Visual C++ memory leak debug tracing
#if defined(_MSC_VER) && defined(_DEBUG)
    _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
#endif

    CppCheckExecutor exec;
#ifdef _WIN32
    GetModuleFileNameA(nullptr, exename, sizeof(exename)/sizeof(exename[0])-1);
    argv[0] = exename;
#endif
```

and as follows on Linux:
```c++
#ifdef __linux__
    // Executing cppcheck in PATH. argv[0] does not contain the path.
    if (mSettings->exename.find_first_of("/\\") == std::string::npos) {
        char buf[PATH_MAX] = {0};
        if (FileLister::fileExists("/proc/self/exe") && readlink("/proc/self/exe", buf, sizeof(buf)-1) > 0)
            mSettings->exename = buf;
    }
#endif
```

So @ericLemanissier is right.
Just don't pas `FILESDIR` to cmake and make sure you put the cfg files in the correct location.

Comment (madebr) [reply recipes/cppcheck/all/conanfile.py:54]: Asssuming every use of a `.cfg` file is using `exename`.

File: recipes/cppcheck/all/conanfile.py [hunk 2]
```diff
@@ -16,8 +17,7 @@ class CppcheckConan(ConanFile):
     options = {"with_z3": [True, False], "have_rules": [True, False]}
     default_options = {"with_z3": True, "have_rules": True}
     exports_sources = ["CMakeLists.txt", "patches/**"]
-
-    _cmake = None
+    build_policy = "always"
```
Comment (SSE4) [inline recipes/cppcheck/all/conanfile.py:20]:
```suggestion
```

File: recipes/cppcheck/all/conanfile.py [hunk 3]
```diff
@@ -44,16 +44,16 @@ def source(self):
...
-        self._cmake.definitions["USE_MATCHCOMPILER"] = "Auto"
-        self._cmake.definitions["ENABLE_OSS_FUZZ"] = False
-        self._cmake.configure(build_folder=self._build_subfolder)
-        return self._cmake
+        cmake = CMake(self)
+        cmake.definitions["USE_Z3"] = self.options.with_z3
+        cmake.definitions["HAVE_RULES"] = self.options.have_rules
+        cmake.definitions["USE_MATCHCOMPILER"] = "Auto"
+        cmake.definitions["ENABLE_OSS_FUZZ"] = False
+        cmake.definitions["FILESDIR"] = os.path.join(self.package_folder, "bin")
```
Comment (SSE4) [inline recipes/cppcheck/all/conanfile.py:54]:
```suggestion
```
seems like it's not required to be set, per @madebr

Comment (madebr) [reply recipes/cppcheck/all/conanfile.py:54]: I give no warranty on any of my claims :D

**Comment by MartinDelille** [comment]
Thank you for your precious feedback! I will patch cppcheck to make it compatible as proposed here as suggested: https://stackoverflow.com/questions/1023306/finding-current-executables-path-without-proc-self-exe/1024937#1024937

**Comment by ericLemanissier** [comment]
my findings: cppcheck [uses GetModuleFileName on windows](https://github.com/danmar/cppcheck/blob/main/cli/main.cpp#L95) and [readlink("/proc/self/exe") on linux](https://github.com/danmar/cppcheck/blob/3feb496197dc1988ad203e6e528e10446dc9c6a7/cli/cmdlineparser.cpp#L133), but it uses plain `argv[0]` on Macos, so you probably just need to add specific macos code calling _NSGetExecutablePath at the beginning of `CmdLineParser::parseFromArgs`

File: recipes/cppcheck/all/conanfile.py (hunk 1 above)
Comment (ericLemanissier) [inline recipes/cppcheck/all/conanfile.py:53]:
```suggestion
```

**Comment by ericLemanissier** [comment]
to [quote myself](https://github.com/conan-io/conan-center-index/pull/10423#issuecomment-1106369173), this does not work for cppcheck <= 2.7.3 on linux, because these release do not have https://github.com/danmar/cppcheck/commit/9ef14d179dfb88cbf5eec54a5b297f0be7a69e79, so the easiest solution is to remove 2.7.3 from config.yml and conandata.yml. Also, you can add patches/0003-handle-exename-on-macos.patch to 2.7.4

**Comment by MartinDelille** [comment]
I wasn't able to make the patch working in 2.7.4 so I removed it also.

**Review by ericLemanissier — APPROVED** [review]

**Review by ericLemanissier — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Comment by ericLemanissier** [comment]
Please don't forget to submit the patch upstream, thanks !

**Pull request merged**